### PR TITLE
[R4R][gasoracle]feat: add uniswap to get token ratio for ETH/MNT; support different token ratio mode

### PR DESCRIPTION
**Pull request by Tri-stone**
# Goals of PR

Core changes:

- add a new source to calculate token ratio for ETH/MNT
- support different token ratio mode

Notes:

- add uniswapV3 as a new source to calculate token ratio
- support setting a default token ratio
- add different token ratio mode
    - mode 1:  `TokenRatioMode(0)`,  use real eth/mnt price to set token ratio
    - mode 2:  `TokenRatioMode(1)`, use eth price to set token ratio
    - mode 3:  `TokenRatioMode(2)`,  use default token ratio
    - use mode 1 as a default mode

Related Issues:

- #1013

### CODE REVIEW

File: gas-oracle/tokenprice/tokenprice.go [hunk 1]
```diff
@@ -9,10 +9,48 @@ import (
...
+	// DefaultETHPrice is default eth_price
+	// If SwitchOneDollarTokenRatio valid, use DefaultETHPrice to set token_ratio to mnt price is 1$
+	DefaultETHPrice = 1800
+
+	// RealTokenRatioMode use eth_price / mnt_price to set token_ratio
+	RealTokenRatioMode = TokenRatioMode(0)
+	// DefaultTokenRatioMode use DefaultTokenRatio to set token_ratio
+	DefaultTokenRatioMode = TokenRatioMode(1)
+	// OneDollarTokenRatioMode use eth_price to set token_ratio, so mnt price is 1$
+	OneDollarTokenRatioMode = TokenRatioMode(2)
```
Comment (wukongcheng) [inline gas-oracle/tokenprice/tokenprice.go:49]: Can it be adjusted to arrange in order of priority using mod?

File: gas-oracle/tokenprice/tokenprice.go [hunk 2]
```diff
@@ -25,31 +63,82 @@ func NewClient(url string, frequency uint64) *Client {
...
+		ethPrice = ethPriceCex
+	}
+
+	switch c.tokenRatioMode {
+	case DefaultTokenRatioMode:
+		// use default eth/mnt price to set token ratio
+		ratio = float64(DefaultTokenRatio)
+	case OneDollarTokenRatioMode:
+		// supposing that mnt is 1 USD, so token_ratio is equals to eth_price
+		ratio = ethPrice
```
Comment (wukongcheng) [inline gas-oracle/tokenprice/tokenprice.go:110]: Which mode is selected by default for use?  And how can the mode be dynamically changed during runtime?

File: gas-oracle/tokenprice/tokenprice_uniswap.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	uniswapV3Quoter *bindings.Uniswapv3Quoter
+}
+
+func newUniswapClient(uniswapURL string) (*uniswapClient, error) {
+	l1Client, err := ethclient.Dial(uniswapURL)
+	if err != nil {
+		return nil, err
+	}
+
+	uniswapV3Quoter, err := bindings.NewUniswapv3Quoter(ContractV3Quoter, l1Client)
```
Comment (wukongcheng) [inline gas-oracle/tokenprice/tokenprice_uniswap.go:37]: How can we upgrade to the new interface after Uniswap upgrades to v4?

**Review by wukongcheng — CHANGES_REQUESTED** [review]

File: gas-oracle/tokenprice/tokenprice_uniswap.go (hunk 3 above)
Comment (Tri-stone) [reply gas-oracle/tokenprice/tokenprice_uniswap.go:37]: > How can we upgrade to the new interface after Uniswap upgrades to v4?

The implementation method of Uniswap v4 is not yet determined. We will look at the solution after it is confirmed later.

File: gas-oracle/tokenprice/tokenprice.go (hunk 2 above)
Comment (Tri-stone) [reply gas-oracle/tokenprice/tokenprice.go:110]: > Which mode is selected by default for use? And how can the mode be dynamically changed during runtime?

- Default mode is `RealTokenRatioMode`(`TokenRatioMode(0)`), which will use real eth_price/mnt_price to set token ratio. I will add a `default` here.
- Gasoracle currently does not support dynamic configuration replacement, and a reboot is required for the configuration to take effect. We will make further modifications in the future to support this feature.

File: gas-oracle/tokenprice/tokenprice.go (hunk 1 above)
Comment (Tri-stone) [reply gas-oracle/tokenprice/tokenprice.go:49]: Modified, priority will show like this:
- RealTokenRatioMode = TokenRatioMode(0)
- OneDollarTokenRatioMode = TokenRatioMode(1)
- DefaultTokenRatioMode = TokenRatioMode(2)

**Review by byteflyfunny — APPROVED** [review]

**Review by guoshijiang — APPROVED** [review]

**Review by wukongcheng — APPROVED** [review]
LGTM

**Pull request merged**